modules/core/src/main/scala/org/tessellation/infrastructure/rewards/EpochRewards.scala

```diff
@@ -0,0 +1,116 @@
+package org.tessellation.infrastructure.rewards
+
+import cats.arrow.FunctionK.liftFunction
+import cats.data._
+import cats.effect.Async
+import cats.effect.std.Random
+import cats.syntax.applicative._
+import cats.syntax.applicativeError._
+import cats.syntax.either._
+import cats.syntax.eq._
+import cats.syntax.flatMap._
+import cats.syntax.functor._
+import cats.syntax.show._
+
+import scala.collection.immutable.{SortedMap, SortedSet}
+
+import org.tessellation.domain.rewards._
+import org.tessellation.ext.refined._
+import org.tessellation.schema.ID.Id
+import org.tessellation.schema.address.Address
+import org.tessellation.schema.balance.Amount
+import org.tessellation.schema.epoch.EpochProgress
+import org.tessellation.schema.transaction.{DAGTransaction, RewardTransaction, TransactionAmount}
+import org.tessellation.schema.{GlobalIncrementalSnapshot, SnapshotOrdinal}
+import org.tessellation.sdk.infrastructure.consensus.trigger.{ConsensusTrigger, EventTrigger, TimeTrigger}
+import org.tessellation.security.signature.Signed
+import org.tessellation.syntax.sortedCollection._
+
+import eu.timepit.refined.auto._
+import eu.timepit.refined.numeric.Positive
+import eu.timepit.refined.refineV
+import io.estatico.newtype.ops._
+
+object EpochRewards {
+  def make[F[_]: Async](
+    rewardsPerEpoch: SortedMap[EpochProgress, Amount],
+    programsDistributor: ProgramsDistributor[Either[ArithmeticException, *]],
+    facilitatorDistributor: FacilitatorDistributor[F]
+  ): EpochRewards[F] = new EpochRewards[F] {
+
+    def distribute(artifact: Signed[GlobalIncrementalSnapshot], trigger: ConsensusTrigger): F[SortedSet[RewardTransaction]] = {
+      val facilitators = artifact.proofs.map(_.id)
+      val transactions = artifact.value.blocks.flatMap(_.block.transactions.toSortedSet).map(_.value)
```

Comment:
Problem is that this will affect the interface of this function - since you can no longer accept only the last snapshot and current trigger - you need to include information about the currently accepted blocks/transactions.

Reply:
Ok, so now the function takes:
- `lastArtifact` (in case someone needs it)
- current transactions
- trigger

The only tricky part is that `fees` are distributed from current ones but minting is based on `proofs` from the previous one. I don't know if we can change that in the artifact creation because we do not have the proof yet.